LoRa: State machine work

hasnainvirk opened the pull request:
### Description
Major contribution of this PR is to rework the state machine for our LoRaWAN stack.
State machine diagrams can be found here [for internal people].
https://armh.sharepoint.com/:f:/r/sites/IoT-Connectivity/Shared%20Documents/LoRa/Mbed%20LoRaWAN%20Stack%20State%20Diagrams/Proposal-NewStateMachine?csf=1&e=PnvvkJ

Other commits either fix a discrepancy or improve integrity of the stack.  

Target -> Mbed-OS 5.9

### Pull request type
    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

Commit:
Immutable Payload from radio

Received data buffer from radio driver should be immutable.

Commit:
Initializing band for default channels

We went through an exercise of adding band information to
any new channel being added. Default channels were looked over.
This commits duly adds missing band information to default channels.

Commit:
Remove useless extraction

Channel plan datastructure already contains channel parameters.
Extraction is not needed.

Commit:
Immutable payload pointer to LoRaMacCommand class

Commit:
Immutable payload pointer in LoRaPHY

Commit:
Adding thread safety

Making our LoRaWAN stack thread safe. If RTOS is not present, locks
don't do anything. ScopedLock is used to automate the lock release on
context expiry.

Commit:
State Machine rework

There had been essentially two state machines running in our stack
which was too cumbersome and was not alligned in any symmetry.

In this work we make sure that:
 * There are no callbacks from the MAC layer to Stack controller layer.
 * Primitives are made local to the mac layer and are presented as
   read-only to the stack controller layer.
 * Interrupt handling and processing is moved to the stack controller layer.
 * Reception is divided into smaller units, seperating handling of Join Accept
   and normal data frames. MIC gets its own unit.
 * Extraction of data and MAC commands from the payload is also being done now in
   its own method.
 * To ensure integrity of the stack, and sanctity of the radio payload, we copy the
   radio payload buffer immediately in the rx interrupt and hoist a flag that prevents
   another interrupt from happening for a short while when we are processing the previous
   packet.
 * If an automatic uplink is on going, we do not send a TX_DONE event to application
   anymore as that is logically incorrect.
 * state_controller() is the central engine for the state machine. To save code space and
   memory, we are not handling each and every state in the state_controller(). Some of the states
   which have no processing to be done, are explicitely set.
 * For all the states who need special processing, seperate methods are added.
 * Class A always run to completion to IDLE and CLass C always runs to completion as RECEIVING.

Commit:
Adding custom channel plan support in AS923

The asia pacific region supports custom channel planning and
downlink channel request. By virtue of a mistake, this information
was missing and hence a custom channel support was not working.
Fixes issue #6783.

kivaisan commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -96,19 +79,55 @@ LoRaWANStack::LoRaWANStack()
-     LoRaMacPrimitives.mcps_indication  = callback(this, &LoRaWANStack::mcps_indication_handler);
-     LoRaMacPrimitives.mlme_confirm     = callback(this, &LoRaWANStack::mlme_confirm_handler);
-     LoRaMacPrimitives.mlme_indication  = callback(this, &LoRaWANStack::mlme_indication_handler);
 }
```
Add initialization of _tx_msg and _rx_payload.

kjbracey replied:
There's really no need to be initialising buffers (or there shouldn't be). But anyway, don't forget you can initialise arrays with `_rx_msg()` in the initialiser list. Don't have to use memset.

hasnainvirk replied:
Added in constructor.

kivaisan commented on features/lorawan/lorastack/mac/LoRaMac.cpp:
```diff
@@ -125,851 +119,698 @@ LoRaMac::LoRaMac()
-
-    mac_primitives = NULL;
-    ev_queue = NULL;
 }
```
Add initialization of _mcps_indication, _mcps_confirmation, _mlme_indication and _mlme_confirmation

kjbracey commented on features/lorawan/system/lorawan_data_structures.h:
```diff
@@ -1216,14 +1092,13 @@ typedef struct {
+    LORAMAC_BUSY                    = 0x00000004,
+    LORAMAC_CONNECTED               = 0x00000008,
+    LORAMAC_USE_OTAA                = 0x00000010,
+    LORAMAC_TX_DONE                 = 0x00000020,
```
Mixing flags and enums is always a bit questionable, and here you don't seem to be using them as flags at all - no combinations? But I see you clearing bits and setting bits.

If this is supposed to actually be an enum of states, then probably a good idea to not make it look like a bitfield - get rid of the values, and stop doing bit manipulations on it.

Or if it is supposed to be bits, get rid of the enum, and actually use a bitfield, as you are always manipulating bits individually, and don't need any tricky masking operations.

Maybe it's a mixture of a state and some flags.

hasnainvirk replied:
It is a bit field, members of the bitfield are identified as an enum. Actual storage is as a 32 bit unsigned bitfield.

kjbracey commented on features/lorawan/LoRaWANStack.h:
```diff
@@ -459,21 +450,44 @@ class LoRaWANStack: private mbed::NonCopyable<LoRaWANStack> {
     loramac_tx_message_t _tx_msg;
     loramac_rx_message_t _rx_msg;
     uint8_t _num_retry;
+    uint32_t _ctrl_flags;
```
Okay, so this isn't the enum? Because the compiler's correctly stopping you doing the bit manipulations, I guess?

hasnainvirk replied:
Flags are now made Macros for clarity.

kjbracey commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -492,322 +363,526 @@ int16_t LoRaWANStack::handle_rx(uint8_t* data, uint16_t length, uint8_t& port, i
+        mlme_indication_handler();
+    }
+
+    memset(_rx_payload, 0, sizeof _rx_payload);
```
Why the memset? memsets are one of the most expensive things you can do - you don't want this sort of thing in data paths. Although I guess this isn't very high throughput...

kjbracey commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -492,322 +363,526 @@ int16_t LoRaWANStack::handle_rx(uint8_t* data, uint16_t length, uint8_t& port, i
+    uint8_t pos = 0;
+    mac_hdr.value = payload[pos++];
+
+    /**
```
I'm still not really grasping why all this MAC functionality is being moved out of the MAC. This totally stops the MAC actually being, well, a MAC. You've now got two "stack" and "MAC" classes which actually form the MAC between them, and I'm not immediately seeing where the dividing line is conceptually.

Maybe there is no clear line, but in that case the MAC class could be eliminated and just merge with this.

This is also less robust and clear than the original MAC code it's replacing - this lost the `switch` and is letting through reserved packets as a result.

hasnainvirk commented:
@kjbracey-arm @kivaisan Please review again.

hasnainvirk commented:
@0xc0170 This needs CI, previous cliapp build failed for some reason.

kjbracey commented on features/lorawan/lorastack/mac/LoRaMac.cpp:
```diff
@@ -126,850 +124,740 @@ LoRaMac::LoRaMac()
 
-        _lora_time.start(_params.timers.mac_state_check_timer, 1);
+        default:
+            tr_error("Unsupported data frame type");
```
Given lack of address filtering, might this be too verbose? You'd print this for every packet you overhear from another node talking to the access point. Maybe it's okay for current testing, and non-class-C.

0xc0170 commented:
I reviewed the log, restarted it

kivaisan commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -41,6 +41,16 @@ SPDX-License-Identifier: BSD-3-Clause
+#define IDLE_FLAG                   0x00000000
+#define TX_ONGOING_FLAG             0x00000001
+#define MSG_RECVD_FLAG              0x00000002
+#define BUSY_FLAG                   0x00000004
```
BUSY_FLAG bit is only set and cleared but never checked so I thinki this should be removed.

Commit:
Doc fix

Structure naming in the docs was wrong.

Commit:
Moving msg flags to lorawan_types.h

Message flags are used in the application so the logical place for
them is in lorawan_types.h and not in lorawan_data_structures.h

Commit:
Datarate bug fix in rx windows configs

While configuring RX parameters for the radio, we need to feed in
rx windows 1 and 2 parameters which are computed when we do the transmission.
We are actually setting the physical value of the data rate rather than
data rate table index and the expectation was to set the data rate index.

Commit:
Removing abort from rx in case of FL discrepency

If the frame length is not what we are expecting, it is
found to be a good practise to actually continue with what we
have received rather than aborting. As we have already demodulated
the packet and RX slots are used up, ther is not so much benefit in
dropping that packet.

hasnainvirk commented:
@kjbracey-arm Can you please review again ?

cmonr commented:
/morph build

AnttiKauppila commented:
@0xc0170 This is ready to be merged

hasnainvirk commented:
@0xc0170 @cmonr Can it be merged now ?
